mineral-ui/mineral-ui
# Flatten Menu data

### Description

Flatten data structure for Menu groups

* Groups are now created similarly to dividers, using `{ group: true, text: 'Group Title' }`
* Renders all menu items, including groups, as flat DOM structure
* The children API for MenuGroups has not changed

_Question:  It would be easy at this time to skip navigating to disabled items in Select/Dropdown menus when using the keyboard.  Is this preferred?_

### Motivation and context

Was investigating performance improvements for Dropdown, Select, using virtualized rendering...

* Rendering menu items in a flat DOM structure is necessary to support virtualized rendering.
* Simpler API for end users, though it is a breaking change for anyone using grouped data
* Simplify Menu rendering code

### Screenshots, videos, or demo, if appropriate

https://menu-flat-data--mineral-ui.netlify.com/

### How to test

1. No changes in functionality
2. Keyboard navigation in Select/Dropdown works as expected
3. Verify documentation/example updates

### Types of changes

- Refactor
- Breaking change (fix or feature that would cause existing functionality to change)

The structure of the data for Menu, Dropdown, Select, and CardTitle components has changed.  Previously, grouped data was provided in a nested data structure.  A flat array of items is now required.  Groups are now created similarly to dividers, using { group: true, text: 'Group Title' }.

The theme export for `menuGroupTitleTheme` has been renamed to `menuGroupTheme`.  The `MenuGroup_margin` theme key has been removed, and the `MenuGroupTitle_paddingTop` adjusted to compensate accordingly.

### Checklist
